src/Lean/Linter/ConstructorAsVariable.lean

```diff
@@ -0,0 +1,79 @@
+ modulo namespaces.
+ -/
+register_option linter.constructorNameAsVariable : Bool := {
+  defValue := true,
+  descr := "enable the linter that warns when bound variable names are constructor names"
+}
+
+/--
+Reports when bound variables' names overlap with constructor names for their type. This is to warn
+especially new users that they have built a pattern that matches anything, rather than one that
+matches a particular constructor. Use `linter.constructorNameAsVariable` to disable.
+-/
+def constructorNameAsVariable : Linter where
+  run cmdStx := do
+    let some cmdStxRange := cmdStx.getRange?
+      | return
+
+    let infoTrees := (← get).infoState.trees.toArray
+    let warnings : IO.Ref (Lean.HashMap String.Range (Syntax × Name × Name)) ← IO.mkRef {}
+
+    for tree in infoTrees do
+      tree.visitM' (preNode := fun ci info _ => do
+        match info with
+        | .ofTermInfo ti =>
+          match ti.expr with
+          | .fvar id .. =>
+            let some range := info.range? | return
+            if (← warnings.get).contains range then return
+            let .original .. := info.stx.getHeadInfo | return
+            if ti.isBinder then
+              -- This is a local variable declaration.
+              let some ldecl := ti.lctx.find? id | return
+              -- Skip declarations which are outside the command syntax range, like `variable`s
+              -- (it would be confusing to lint these), or those which are macro-generated
+              if !cmdStxRange.contains range.start || ldecl.userName.hasMacroScopes then return
+              let opts := ci.options
+              -- we have to check for the option again here because it can be set locally
+              if !linter.constructorNameAsVariable.get opts then return
+              if let n@(.str .anonymous s) := info.stx.getId then
+                -- Check whether the type is an inductive type, and get its constructors
+                let ty ←
+                  if let some t := ti.expectedType? then pure t
+                  else ti.runMetaM ci (Meta.inferType ti.expr)
+                let ty ← ti.runMetaM ci (instantiateMVars ty >>= Meta.whnf)
+                if let .const tn _ := ty.getAppFn' then
+                  if let some (.inductInfo i) := (← getEnv).find? tn then
+                    for c in i.ctors do
+                      if let .str _ cn := c then
+                        if cn == s then
+                          warnings.modify (·.insert range (info.stx, n, c))
+            else pure ()
+          | _ => pure ()
+        | _ => pure ())
+
+    -- Sort the outputs by position
+    for (_range, declStx, userName, ctorName) in (← warnings.get).toArray.qsort (·.1.start < ·.1.start) do
+      logLint linter.constructorNameAsVariable declStx <|
+        m!"Local variable '{userName}' resembles constructor '{ctorName}' - " ++
+        m!"missing namespace qualifier or 'open' statement?"
```

Comment:
How about 
```suggestion
        m!"did you want to write .{userName} (with a dot)?"
```
or so? This gives the (likely confused and frustrated) user some immediate and likely to work fix without requiring them to go off and read about possibly new concepts (`open`), and nudges then towards the recommended (at least by y Leo style). Could even become a quick fix later…

Reply:
I think this is a clear improvement, thanks! And the bit about the namespace qualifier seems redundant with the explicit namespace written in the error message.

Reply:
I want to tweak the wording a bit - many of the target users didn't **want** to write `.name`, because this is their first encounter with the syntax, and I don't want the text to seem presumptuous.

Reply:
New wording:
> warning: Local variable 'y' resembles constructor 'A.y' - write '.y' (with a dot) or 'A.y' to use the constructor.